extra/help/lint/coverage/coverage.factor

```diff
@@ -0,0 +1,146 @@
+
+: write-object-seq ( object-seq -- )
+    [
+        dup array? [
+            dup ?first array?
+            [ [ first2 write-object " " write ] each ]
+            [ first2 write-object ] if
+        ] [ write ] if
+    ] each ; inline
+
+: (assemble-word-metadata) ( vec word -- vec )
+    [
+        [ "[" ] dip vocabulary>> dup lookup-vocab 2array "] "
+            3array over push-all
+    ] [
+        [ name>> ] keep 2array ": "
+        2array over push-all
+    ] bi ; inline
+
+: (assemble-empty-examples) ( vec coverage -- vec )
+    empty-examples?>> [ "empty " \ $examples [ name>> ] keep 2array "; "
+        3array over push-all
+    ] when ;
+
+: (assemble-omitted-sections) ( vec coverage -- vec )
+    omitted-sections>> [
+        length "section" ?pluralize ": " append
+    ] [
+        [ [ name>> ] keep 2array ] map
+    ] bi
+    [ "needs help " ] 2dip
+    3array over push-all ;
+
+: (present-coverage) ( coverage-report -- )
+    [ V{ } clone ] dip
+    [ word-name>> (assemble-word-metadata) ] keep
+    dup 100%-coverage?>> [ drop "full help coverage" over push ]
+    [ [ (assemble-empty-examples) ] [ (assemble-omitted-sections) ] bi ] if
+    "\n" over push
+    write-object-seq ; inline
+
+M: word-help-coverage summary
+    [ (present-coverage) ] with-string-writer ; inline
+
+: sorted-loaded-child-vocabs ( prefix -- assoc )
+    loaded-child-vocab-names natural-sort ; inline
+
+: resolve-name-in ( name namespaces -- word )
+    "syntax" swap remove " " join
+    "USING: " " ; \\ " surround
+    prepend eval( -- word ) ;
+
+: filter-private ( seq -- no-private )
+    [ ".private" ?tail nip not ] filter ; inline
+
+: ?pluralize ( n singular -- singular/plural )
+    count-of-things " " split1 nip ;
+
+: should-define ( word -- spec )
+    {
+        { [ dup predicate? ]    [ drop { } ] } ! predicate?s have generated docs
+        { [ dup error-class? ]  [ drop { $values $description $error-description } ] }
+        { [ dup class? ]        [ drop { $class-description } ] }
+        { [ dup generic? ]      [ drop { $values $contract $examples } ] }
+        { [ dup word? ]         [ drop { $values $description $examples } ] }
+        [ drop no-cond ]
+    } cond ;
+
+: word-defines-sections ( word -- seq )
+    word-help [ ignored-words member? not ] filter [ ?first ] map ;
+
+! only words that need examples, need to have them nonempty
+! not defining examples is not the same as an empty { $examples }
+: empty-examples? ( word -- ? )
+    word-help \ $examples swap elements [ f ] [ first rest empty? ] if-empty ;
+
+: missing-sections ( word -- missing )
+    [ should-define ] [ word-defines-sections ] bi diff ;
+PRIVATE>
+
+GENERIC: <word-help-coverage> ( word -- coverage )
+M: word <word-help-coverage>
+    dup [ missing-sections ] [ empty-examples? ] bi
+    2dup 2array { { } f } =
+    word-help-coverage boa ; inline
+
+M: string <word-help-coverage>
+    loaded-vocab-names resolve-name-in call-next-method ; inline
```

Comment:
Not a big fan of this approach to looking up words from strings, I think maybe better to either not support string resolution of names, or do something like:

```factor
: find-word ( name -- word/f )
    dup dictionary get values [ lookup-word ] with map sift
    dup length {
        { 0 [ 2drop f ] }
        { 1 [ first nip ] }
        [ drop <ambiguous-use-error> throw-restarts ]
    } case ;
```

So if you try, for example, ``"search" find-word`` you get a restart with the three possibilities in binary-search, peg.search, and, vocabs.parser.

Reply:
There is also a `find-word`  already existing in `fuel.help`

Reply:
Hmm... the use of `loaded-vocab-names resolve-name-in` was that the string would be looked up in the exact same context where it was written, and `fuel.help:find-word` seems good for that purpose, so I'll use that

Reply:
Looks like when the test runner runs ``help.lint.coverage`` tests, the ``"eye" find-word`` fails because ``search`` (called by ``fuel.help:find-word``) throws an error that ``manifest`` is ``f``...

```
==== resource:extra/help/lint/coverage/coverage-tests.factor

resource:extra/help/lint/coverage/coverage-tests.factor: 37

Unit Test: { { eye } [ "eye" find-word ] }

Generic word qualified-vocabs>> does not define a method for the POSTPONE: f class.
Dispatching on object: f
```

Since the listener has ``manifest`` set to the ``scratchpad`` vocabulary (causing the tests to pass when run manually), you can reproduce this way:

```factor
IN: scratchpad f manifest [ "help.lint.coverage" test ] with-variable
```

For the moment, I put a ``find-word`` that is more like the one I suggested, so the tests pass, but feel free to either just call ``words-named ?first`` or change it in some other way.  I only patched it to fix the build not to overwrite your ideas.